.github/workflows/validate-commit-message.py

```diff
@@ -2,82 +2,153 @@
-        if len(commit.parents) > 1:
-            print('Not going beyond a merge commit')
-            break
-
-        prefix = commit.message.splitlines()[0]
-        if prefix_invalid(prefix):
-            quoted_commit_message = textwrap.indent(commit.message, prefix="  > ")
-            msg = f'* {commit.hexsha}\n{quoted_commit_message}'
-            faulty_commits.append(msg)
-
-    if faulty_commits:
-        warning = 'These commits are not formatted correctly. '
-        warning += 'Please amend them to start with one of:\n'
-        warning += '* \\<package>: \n'
-        warning += '* \\<package>, <package>, ...: \n'
-        warning += f'* {", ".join(keyword + ":" for keyword in KEYWORDS)}\n\n'
-        warning += "### Faulty commits:\n"
-        faulty_commits.insert(0, warning)
-        with open('faulty_commits.txt', 'w') as fp:
-            fp.write('\n'.join(faulty_commits))
-        with open(os.environ['GITHUB_OUTPUT'], 'a') as fp:
-            fp.write("faulty-commits=true")
-
-
-if __name__ == '__main__':
+def get_changed_packages(changed_files: list[str]) -> list[str]:
+    """
+    Return all packages changed by the commit
+    """
+
+    changed_packages = []
+    changed_package_paths = [path for path in changed_files if "/packages/" in path]
+    for package_path in changed_package_paths:
+        path_components = package_path.split("/")
+        changed_packages.append(
+            path_components[path_components.index("packages") + 1]
+        )
+
+    return changed_packages
+
+
+def get_unmentioned_packages(prefixes: list[str], changed_files: list[str]) -> list[str]:
+    unmentioned_packages = []
+
+    changed_packages = get_changed_packages(changed_files)
+
+    for package in changed_packages:
+        if package not in prefixes:
+            unmentioned_packages.append(package)
+
+    return unmentioned_packages
+
+
+def collect_prefixes(message: str) -> list[str]:
+    """
+    Collect all prefixes in the commit message
+    """
+    prefixes = []
+
+    for line in message.splitlines():
+        if ":" in line:
+            prefix = message.splitlines()[0]
+            prefix_items = [item.strip() for item in prefix.split(",")]
+            prefixes.extend(prefix_items)
+
+    return prefixes
+
+
+def main(title: str, changed_files: list[str]) -> None:
+    print(
+        "Setting fail state to make sure we catch any script failures- we'll clean up at the end"
+    )
+    with open(os.environ["GITHUB_OUTPUT"], "a") as fp:
+        fp.write("faulty-commits=true")
+    repo = Repo(".")
+
+    commit_message_issues = []
+
+    commit = next(repo.iter_commits())
+    print(f"Checking commit: {commit.message} (parents: {commit.parents})")
+    quoted_commit_message = textwrap.indent(commit.message, prefix="  > ")
+    prefixes = collect_prefixes()
+
+    unmentioned_packages = get_unmentioned_packages(prefixes, changed_files)
+    if unmentioned_packages:
```

Comment:
Devil's advocate: if we mass-update the copyright, this is going to blow up :D

Reply:
Very good point - maybe we should temporarily turn this off for that kind of changes 🥷

Reply:
Or define a custom threshold like 3 or 5 packages changed total?